Cabal/Distribution/Package.hs

```diff
@@ -77,12 +116,16 @@ import Text.PrettyPrint ((<+>), text)
 -- This type is opaque since @Cabal-2.0@
 --
 -- @since 2.0
-newtype PackageName = PackageName ShortText
-    deriving (Generic, Read, Show, Eq, Ord, Typeable, Data)
+newtype PackageName = PackageName UnqualComponentName
+    deriving (Generic, Read, Show, Eq, Ord, Typeable, Data,
+              Binary, Text, NFData)
+
+packageNameToUnqualComponentName :: PackageName -> UnqualComponentName
```

Comment:
Needs doc. Something like "Converts a package name to an unqualified component name; useful in legacy situations where a package name may refer to an internal component, if one is defined with that name."